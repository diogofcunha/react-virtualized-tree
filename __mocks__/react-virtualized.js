import React from 'react';

export class CellMeasurerCache {}

export const AutoSizer = ({children}) => <React.Fragment>{children({width: 1000, height: 1000})}</React.Fragment>;
// const mockInifiteLoader = ({children}) => (
//   <React.Fragment>{children({registerChild: jest.fn(), onRowsRendered: jest.fn()})}</React.Fragment>
// );

export const CellMeasurer = ({children}) => <React.Fragment>{children({measure: jest.fn()})}</React.Fragment>;

export const List = p =>
  Array.from({length: p.rowCount}).map((_, i) =>
    p.rowRenderer({
      key: String(i),
      columnIndex: 0,
      index: i,
      rowIndex: i,
      isScrolling: false,
      isVisible: true,
      parent: {},
      style: {},
    }),
  );
