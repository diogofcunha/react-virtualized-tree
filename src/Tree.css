.tree-node {
  user-select: none;
  cursor: pointer;
}
