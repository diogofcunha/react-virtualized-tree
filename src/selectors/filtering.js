const INITIAL_FILTERED_VALUE = {nodes: [], nodeParentMappings: {}};

export const filterNodes = (filter, nodes, parents = []) =>
  nodes.reduce((filtered, n) => {
    const {nodes: filteredChildren, nodeParentMappings: childrenNodeMappings} = n.children
      ? filterNodes(filter, n.children, [...parents, n.id])
      : INITIAL_FILTERED_VALUE;

    return !(filter(n) || filteredChildren.length)
      ? filtered
      : {
          nodes: [
            ...filtered.nodes,
            {
              ...n,
              children: filteredChildren,
            },
          ],
          nodeParentMappings: {
            ...filtered.nodeParentMappings,
            ...childrenNodeMappings,
            [n.id]: parents,
          },
        };
  }, INITIAL_FILTERED_VALUE);
