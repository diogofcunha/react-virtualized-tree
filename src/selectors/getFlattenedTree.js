export const isNodeExpanded = node => node.state && node.state.expanded;
export const nodeHasChildren = node => node.children && node.children.length;

export const getFlattenedTree = (nodes, parents = []) =>
  nodes.reduce((flattenedTree, node) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents};

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(node.children, [...parents, node.id])];
  }, []);

export const getFlattenedTreePaths = (nodes, parents = []) => {
  const paths = [];

  for (const node of nodes) {
    const {id} = node;

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      paths.push(parents.concat(id));
    } else {
      paths.push(parents.concat(id));
      paths.push(...getFlattenedTreePaths(node.children, [...parents, id]));
    }
  }

  return paths;
};

export const doesChangeAffectFlattenedTree = (previousNode, nextNode) => {
  return isNodeExpanded(previousNode) !== isNodeExpanded(nextNode);
};
