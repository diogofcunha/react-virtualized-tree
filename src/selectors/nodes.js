import { createSelector } from 'reselect';
import { omit } from 'lodash';

import { UPDATE_TYPE } from '../contants';

export const getNodeRenderOptions = createSelector(
  (node => (node.state || {}).expanded),
  (node => (node.state || {}).favorite),
  (node => (node.state || {}).deletable),
  (node => node.children), 
  (expanded, favorite, deletable, children = []) => ({
    hasChildren: !!children.length,
    isExpanded: !!expanded,
    isFavorite: !!favorite,
    isDeletable: !!deletable
  })
);

const FLATTEN_TREE_PROPERTIES = [ 'deepness', 'parents' ];
const NODE_OPERATION_PROPERTIES = [ 'index' ];

const NODE_OPERATION_TYPES = {
  CHANGE_NODE: 'CHANGE_NODE',
  DELETE_NODE: 'DELETE_NODE',
  ADD_NODE: 'ADD_NODE'
}

const NODE_CHANGE_OPERATIONS = {
  CHANGE_NODE: (nodes, updatedNode) => nodes.map(n => n.id === updatedNode.id ? omit(updatedNode, FLATTEN_TREE_PROPERTIES) : n),
  DELETE_NODE: (nodes, updatedNode) => nodes.filter(n => n.id !== updatedNode.id),
  ADD_NODE: (nodes, nodeToAdd) => [ ...nodes, omit(nodeToAdd, [ ...FLATTEN_TREE_PROPERTIES, ...NODE_OPERATION_PROPERTIES] ) ]
}

export const replaceNodeFromTree = (nodes, updatedNode, operation = NODE_OPERATION_TYPES.CHANGE_NODE) => {
  if (!NODE_CHANGE_OPERATIONS[operation]) {
    return nodes;
  }
  
  const { parents } = updatedNode;

  if (!parents.length) {
    return NODE_CHANGE_OPERATIONS[operation](nodes, updatedNode);
  }

  const parentIndex = nodes.findIndex(n => n.id === parents[0])

  const preSiblings = nodes.slice(0, parentIndex);
  const postSiblings = nodes.slice(parentIndex + 1);

  return [
    ...preSiblings,
    {
      ...nodes[parentIndex], 
      ...nodes[parentIndex].children || operation === NODE_OPERATION_TYPES.ADD_NODE ? 
        { children: replaceNodeFromTree(
            nodes[parentIndex].children || [],
            { ...updatedNode, parents: parents.slice(1) },
            operation
          )
        } : {}
    },
    ...postSiblings
  ]
}

export const deleteNodeFromTree = (nodes, deletedNode) => {
  return replaceNodeFromTree(
    nodes,
    deletedNode,
    NODE_OPERATION_TYPES.DELETE_NODE
  );
}

export const addNodeToTree = (nodes, nodeToAdd) => replaceNodeFromTree(
  nodes,
  nodeToAdd,
  NODE_OPERATION_TYPES.ADD_NODE
)

export const moveNodeToChildren = (nodes, [ fromNode, toNode ]) => {
  const { id: targetId, parents: targetParents } = toNode;
  const parents = [ ...targetParents, ...targetId !== fromNode.id ? [targetId] : []];

  return moveNodeFromTree(nodes, [ fromNode, { ...toNode, parents } ]);
}

export const moveNodeFromTree = (nodes, [ fromNode, toNode ]) => {
  const treeWithoutMovedNode = deleteNodeFromTree(
    nodes,
    fromNode
  );

  return addNodeToTree(treeWithoutMovedNode, { ...fromNode, parents: toNode.parents });
}

export const udpateNode = (originalNode, newState) => ({
  node: {
    ...originalNode,
    state : {
      ...originalNode.state,
      ...newState
    }
  },
  type: UPDATE_TYPE.UPDATE
});

export const deleteNode = node => ({
  node,
  type: UPDATE_TYPE.DELETE
});

export const addNode = node => ({
  node,
  type: UPDATE_TYPE.ADD
});
