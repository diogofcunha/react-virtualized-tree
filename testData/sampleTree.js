export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
