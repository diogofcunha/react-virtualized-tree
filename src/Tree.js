import React from 'react';
import PropTypes from 'prop-types';
import { AutoSizer, List, CellMeasurerCache, CellMeasurer } from 'react-virtualized';

import { FlattenedNode } from './shapes/nodeShapes';

export default class Tree extends React.Component {
  _cache = new CellMeasurerCache({
    fixedWidth: true,
    minHeight: 20,
  })

  rowRenderer = ({ node, key, measure, style, NodeRenderer }) => {
    const {nodeMarginLeft} = this.props;

    return (
      <div
        key={key}
        className="tree-node"
        style={{ ...style, marginLeft: node.deepness * nodeMarginLeft}}
      >
        <NodeRenderer node={node} onChange={this.props.onChange} measure={measure}/>
      </div>
    );
  }

  measureRowRenderer =  (nodes) => ({ key, index, style, parent }) => {
    const { NodeRenderer } = this.props;
    const node = nodes[index];

    return (
      <CellMeasurer
        cache={this._cache}
        columnIndex={0}
        key={key}
        rowIndex={index}
        parent={parent}>
        {
          m => this.rowRenderer({ ...m, node, key, style, NodeRenderer })
        }
      </CellMeasurer>
    )
  }

  render() {
    const { nodes, rowHeight } = this.props;

    return (
      <AutoSizer>
        {({ height, width }) => (
          <List
            deferredMeasurementCache={this._cache}
            ref={r => this._list = r}
            height={height}
            rowCount={nodes.length}
            rowHeight={rowHeight || this._cache.rowHeight}
            rowRenderer={this.measureRowRenderer(nodes)}
            width={width}
          />
        )}
      </AutoSizer>
    );
  }
}

Tree.propTypes = {
  nodes: PropTypes.arrayOf(PropTypes.shape(FlattenedNode)).isRequired,
  NodeRenderer: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  nodeMarginLeft: PropTypes.number,
  rowHeight: PropTypes.number,
};
