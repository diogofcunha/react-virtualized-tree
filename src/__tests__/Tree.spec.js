import React from 'react';
import { shallow } from 'enzyme';
import { List } from 'react-virtualized';

import Tree from '../Tree';
import { Nodes } from '../../testData/sampleTree';
import { getFlattenedTree } from '../selectors/getFlattenedTree';

describe('Tree', () => {
  const SampleRenderer = ({ node }) => (
    <div>
      {`${node.id}`}
    </div>
  );

  const nodes = getFlattenedTree(Nodes);

  const setup = () => shallow(
    <Tree nodes={nodes} onChange={jest.fn()}>
      {
        ({ node, ...rest }) =>
          <SampleRenderer
            node={node}
          >
            { node.name }
          </SampleRenderer>
      }
    </Tree>
  );

  it('should render a Tree with children and the correct props', () => {    
    const wrapper = setup();

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('rowRenderer should render the Node renderer with the correct props', () => {
    const wrapper = setup();

    const { rowRenderer: rowRendererCb } = wrapper.instance();
    const RowRenderer = rowRendererCb(nodes);
    const rowRendererProps = { key: 'a', index: 0, style: {} };
    
    const rowRendererWrapper = shallow(<RowRenderer {...rowRendererProps}/>);

    expect(rowRendererWrapper.first().children().props()).toMatchSnapshot();
  });
});
