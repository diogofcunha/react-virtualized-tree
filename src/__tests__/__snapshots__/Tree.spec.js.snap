// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree measureRowRenderer should render a CellMeasurer 1`] = `
Object {
  "cache": CellMeasurerCache {
    "_cellHeightCache": Object {},
    "_cellWidthCache": Object {},
    "_columnCount": 0,
    "_columnWidthCache": Object {},
    "_defaultHeight": 30,
    "_defaultWidth": 100,
    "_hasFixedHeight": false,
    "_hasFixedWidth": true,
    "_keyMapper": [Function],
    "_minHeight": 20,
    "_minWidth": 0,
    "_rowCount": 0,
    "_rowHeightCache": Object {},
    "columnWidth": [Function],
    "rowHeight": [Function],
  },
  "children": [Function],
  "columnIndex": 0,
  "parent": Object {},
  "rowIndex": 0,
}
`;

exports[`Tree rowRenderer should render the Node renderer with the correct props 1`] = `
Object {
  "measure": [MockFunction],
  "node": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": true,
    },
  },
  "onChange": [MockFunction],
}
`;

exports[`Tree should render a Tree with children and the correct props 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "overflow:visible;height:0;width:0",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "aria-label": "grid",
          "aria-readonly": "true",
          "class": "ReactVirtualized__Grid ReactVirtualized__List",
          "role": "grid",
          "style": "box-sizing:border-box;direction:ltr;height:0;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:auto",
          "tabindex": "0",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-label": undefined,
          "aria-readonly": undefined,
          "class": undefined,
          "role": undefined,
          "style": undefined,
          "tabindex": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-label": undefined,
          "aria-readonly": undefined,
          "class": undefined,
          "role": undefined,
          "style": undefined,
          "tabindex": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
